### ISSUE TITLE
val_loss not available (validation set and --compute-val-loss are provided)

### ISSUE THREAD

**Issue by mrtgenet** [fizyr/keras-retinanet#976]
Hi,

**First thanks for the helpful tools you provide here ! Really appreciate :)**

Il searched a bit through open and closed issues but I didn't manage to find answer, though some issues are related. So let me quickly explain : 

I am training a net with train and val sets, all works fine with default settings and I have promising results. But I would like to check validation loss instead of training loss (saving only best snapshots), which should be possible by changing the monitor value to 'val_loss' in the ModelCheckPoint callback and providing the "--compute-val-loss" argument. 

But when I tried, I got the following error : 
**"RuntimeWarning: Can save best model only with val_loss available, skipping.
  'skipping.' % (self.monitor), RuntimeWarning)"**

Here the corresponding code section in train.py : 
```
# save the model
if args.snapshots:
  # ensure directory created first; otherwise h5py will error after epoch.
  makedirs(args.snapshot_path)
  checkpoint = keras.callbacks.ModelCheckpoint(
    os.path.join(
      args.snapshot_path,
      '{backbone}_{dataset_type}_{{epoch:02d}}.h5'.format(backbone=args.backbone, dataset_type=args.dataset_type)
    ),
    verbose=1,
    save_best_only=True,
    monitor="val_loss",
    mode='min'
)
checkpoint = RedirectModel(checkpoint, model)
callbacks.append(checkpoint)
```

And here is the command I run  : 
```
keras_retinanet/bin/train.py
  --weights snapshots/ResNet-50-model.keras.h5 
  --batch-size 12 
  --epochs 90 
  --steps 1000 
  --freeze-backbone 
  --random-transform 
  --config keras_retinanet/config.ini 
  --compute-val-loss 
  csv path/to/train_set.csv path/to/class_map.csv --val-annotations path/to/val_set.csv
```

Is there something wrong or do I misunderstand the "--compute-val-loss" argument ?

Also I checked the code and I found that "--compute-val-loss" action in arg parser is 'store_false', is it normal ? It seems weird at first glance, though I must tell I didn't looked through the whole code to see how exactly it is used.

Thank you for any tip, 
Martin

**Comment by liminghuiv** [comment]
Hi @mrtgenet ,
I had similar issue before. '--compute-val-loss' is not available. I would suggest that we add it, it is useful.

**Comment by de-vri-es** [comment]
> Also I checked the code and I found that "--compute-val-loss" action in arg parser is 'store_false', is it normal ? It seems weird at first glance, though I must tell I didn't looked through the whole code to see how exactly it is used.

That does sound weird. Judging from the way it's used, it should be `store_true`:
https://github.com/fizyr/keras-retinanet/blob/3b00e7ef19692b473b25df69e5a5e555c6fcb48b/keras_retinanet/bin/train.py#L496-L497

@mrtgenet: Do you feel like submitting a PR?

**Comment by mrtgenet** [comment]
@de-vri-es I didn't take the time to test modifying this stuff until now 'cause I have different models training right now (testing different hyper parameters) but I will try and see. But it seems weird to me that such a simple true/false thing would be an error, consider the number of poeple using that framework, and not a (weird) choice that made things more simple on the moment idk. That's why I prefered asking here first.

**Comment by mrtgenet** [comment]
Well ok sometimes things are just that simple lol
Simple test with 
`parser.add_argument('--compute-val-loss', help='Compute validation loss during training', dest='compute_val_loss', action='store_true')`

and I get the validation losses as expected with argument  "--compute-val-loss" given to the _train.py_ call:
```
Epoch 1/50
100/100 [==============================] - 147s 1s/step - loss: 2.5200 - regression_loss: 1.7152 - classification_loss: 0.8048 - val_loss: 2.0127 - val_regression_loss: 1.3301 - val_classification_loss: 0.6826
```

@liminghuiv here is your solution :)

I don't know if this deserves a PR

**Comment by hgaiser** [comment]
That argument is a recent contribution, must've slipped through the review. It does seem like an error, indeed. Even though it is such a small change, could you make a PR?

**Comment by mrtgenet** [comment]
I work in an unadapted environment (github is kinda blocked, I used zip archive to download) so I can't right now, but I could do this at home I guess. Or is there a way to do this on navigator ?

**Comment by mrtgenet** [comment]
Ah whatever it's simpler than I thought haha, never did it that way I didn't know :^)

**Comment by de-vri-es** [comment]
Thanks for the PR. Considering the're a fix at #978, I think we can close this issue.

Thanks!

**Comment by mrtgenet** [comment]
np :)

**Comment by mariaculman18** [comment]
I am working with the last version (0.5.1) and `--compute-val-loss` is either ignored or not recognize. Any clue what happened?

**Comment by kristofgiber** [comment]
@de-vri-es When is this fix going to be available in TF2-RC or beta? I've updated both to the latest version but still getting the above error when trying to train with val_loss (and validation_split provided)

**Comment by hgaiser** [comment]
> @de-vri-es When is this fix going to be available in TF2-RC or beta? I've updated both to the latest version but still getting the above error when trying to train with val_loss (and validation_split provided)

What error do you mean exactly? There was only a warning stating that it can't monitor `val_loss` if it is not created. This should be independent from using TF2-RC, beta or TF 1.14.
